tests/test_quantumBackedButler.py

```diff
@@ -0,0 +1,333 @@
+        self.datasetTypeOutput = DatasetType("test_ds_output", graph, storageClass)
+        self.datasetTypeOutput2 = DatasetType("test_ds_output2", graph, storageClass)
+        self.datasetTypeExtra = DatasetType("test_ds_extra", graph, storageClass)
+        self.butler.registry.registerDatasetType(self.datasetTypeInit)
+        self.butler.registry.registerDatasetType(self.datasetTypeInput)
+        self.butler.registry.registerDatasetType(self.datasetTypeOutput)
+        self.butler.registry.registerDatasetType(self.datasetTypeOutput2)
+        self.butler.registry.registerDatasetType(self.datasetTypeExtra)
+
+        dataIds = [
+            self.butler.registry.expandDataId(dict(instrument="Cam1", detector=detector_id))
+            for detector_id in (1, 2, 3, 4)
+        ]
+
+        # make actual input datasets
+        self.input_refs = [
+            self.butler.put({"data": dataId["detector"]}, self.datasetTypeInput, dataId) for dataId in dataIds
+        ]
+        self.init_inputs_refs = [self.butler.put({"data": -1}, self.datasetTypeInit, dataIds[0])]
+
+        # generate dataset refs for outputs
+        self.output_refs = [
+            DatasetRef(self.datasetTypeOutput, dataId, id=uuid.uuid4(), run="RUN") for dataId in dataIds
+        ]
+        self.output_refs2 = [
+            DatasetRef(self.datasetTypeOutput2, dataId, id=uuid.uuid4(), run="RUN") for dataId in dataIds
+        ]
+
+        self.missing_refs = [
+            DatasetRef(self.datasetTypeExtra, dataId, id=uuid.uuid4(), run="RUN") for dataId in dataIds
+        ]
+
+    def tearDown(self):
+        removeTestTempDir(self.root)
+
+    def make_quantum(self, step: int = 1) -> Quantum:
+        """Make a Quantum which includes datastore records."""
+
+        if step == 1:
+            datastore_records = self.butler.datastore.export_records(self.input_refs + self.init_inputs_refs)
+            predictedInputs = {self.datasetTypeInput: self.input_refs}
+            outputs = {self.datasetTypeOutput: self.output_refs}
+            initInputs = {self.datasetTypeInit: self.init_inputs_refs}
+        elif step == 2:
+            # The result should be empty, this is just to test that it works.
+            datastore_records = self.butler.datastore.export_records(self.output_refs)
+            predictedInputs = {self.datasetTypeInput: self.output_refs}
+            outputs = {self.datasetTypeOutput2: self.output_refs2}
+            initInputs = {}
+
+        return Quantum(
+            taskName="some.task.name",
+            inputs=predictedInputs,
+            outputs=outputs,
+            initInputs=initInputs,
+            datastore_records=datastore_records,
+        )
+
+    def test_initialize(self):
+        """Test for initialize method"""
+
+        quantum = self.make_quantum()
+        qbb = QuantumBackedButler.initialize(config=self.config, quantum=quantum, dimensions=self.universe)
```

Comment:
I had naively assumed that people would still use `Butler(some_thing)` to instantiate a `QuantumBackedButler` -- point at a config file that includes a pointer to the graph file. Did we decide that won't be useful because no-one outside execution is going to be wanting to create one of these and we are in charge of all the code that would?

Reply:
My impression was indeed that this is quite special and not generally useful outside execution. Besides, QBB is a LimitedButler, not a full Butler.

Reply:
`LimitedButler.from_config()` may make sense to avoid special casing in ctrl_mpexec (if it can be done at all). I'll figure it out when I start working on that.

Reply:
I had punted on the question of construction entirely when thinking about this before.  I'd be inclined to use a `QuantumBackedButler` constructor or classmethod directly instead of trying to come up with an interface that works for any `LimitedButler`, as they could have very different arguments that wouldn't necessarily coexist well in a single signature.

Reply:
In my world the calling interface for `LimitedButler.from_config()` would be a URI and that URI would be a config file that has a `cls` field that allows the correct butler class to be imported which would then be given the config as normal. Things like collections and dataId args would be the same.

Reply:
> Things like collections and dataId args would be the same.

I think this makes my point; a `QuantumBackedButler` doesn't get to pick its collections or its default data IDs, because those aren't meaningful concepts for a `LimitedButler`.  Instead, it needs to be passed a `Quantum` or a way to get read `Quantum`.

Reply:
I was putting the URI of the quantum graph in the YAML file...

Reply:
Ok, fair, so we _can_ clearly pass in anything any `LimitedButler` needs via a config file, but what do we gain from doing that?  I don't think we have a use case for a consistent construction interface, just a consistent get/put/etc. use case.  And it still seems like we lose by having arguments in the signature that end up getting ignored by most implementations.

Reply:
In this specific case where a general user is not going to be confused it's fine. I think we are currently fine with "remote butler" as well because that still has to be a full butler and it's the registry class that gets replaced not the butler class. If remote butler becomes something that is its own butler class then we will have to probably have Butler be able to do something clever based on the content of the configuration YAML so that it can switch itself into a RemoteButler vs LocalButler without requiring everyone to change their constructor calls from `Butler(config)` to `Butler.from_config(config)`.